code/modules/power/pt_laser.dm

```diff
@@ -339,158 +337,96 @@
-		if(1e9)
-			t += "Input level: <A href = '?src=\ref[src];set_input=1'>[input_number]</A> <A href = '?src=\ref[src];input=0'>W</A> <A href = '?src=\ref[src];input=1'>kW</A> <A href = '?src=\ref[src];input=2'>MW</A> <B>GW</B> <A href = '?src=\ref[src];input=4'>TW</A><BR>"
-		if(1e12)
-			t += "Input level: <A href = '?src=\ref[src];set_input=1'>[input_number]</A> <A href = '?src=\ref[src];input=0'>W</A> <A href = '?src=\ref[src];input=1'>kW</A> <A href = '?src=\ref[src];input=2'>MW</A> <A href = '?src=\ref[src];input=3'>GW</A> <B>TW</B><BR>"
-
-	t += "<BR><BR>"
-
-	t += "Output: [online ? "<B>Online</B> <A href = '?src=\ref[src];online=1'>Offline</A>" : "<A href = '?src=\ref[src];online=1'>Online</A> <B>Offline</B> "]<BR>"
-
-	switch(output_multi)
-		if(1e6)
-			t += "Output level: <A href = '?src=\ref[src];set_output=1'>[output_number]</A> <B>MW</B> <A href = '?src=\ref[src];output=3'>GW</A> <A href = '?src=\ref[src];output=4'>TW</A><BR>"
-		if(1e9)
-			t += "Output level: <A href = '?src=\ref[src];set_output=1'>[output_number]</A> <A href = '?src=\ref[src];output=2'>MW</A> <B>GW</B> <A href = '?src=\ref[src];output=4'>TW</A><BR>"
-		if(1e12)
-			t += "Output level: <A href = '?src=\ref[src];set_output=1'>[output_number]</A> <A href = '?src=\ref[src];output=2'>MW</A> <A href = '?src=\ref[src];output=3'>GW</A> <B>TW</B><BR>"
-
-	t += "<BR><br>lifetime earnings:<br>[engineering_notation(lifetime_earnings)] credits</PRE><HR><A href='?src=\ref[src];close=1'>Close</A>"
-
-	t += "</TT>"
-	user.Browse(t, "window=Power Transmission Laser;size=460x300")
-	onclose(user, "Power Transmission Laser")
-	return
-
-/obj/machinery/power/pt_laser/Topic(href, href_list)
-	..()
-
-	if (usr.stat || usr.restrained() )
+/obj/machinery/power/pt_laser/ui_state(mob/user)
+	return tgui_default_state
+
+/obj/machinery/power/pt_laser/ui_status(mob/user, datum/ui_state/state)
+	return min(
+		state.can_use_topic(src, user),
+		tgui_broken_state.can_use_topic(src, user),
+		tgui_not_incapacitated_state.can_use_topic(src, user)
+	)
+
+/obj/machinery/power/pt_laser/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "PowerTransmissionLaser")
+		ui.open()
+
+/obj/machinery/power/pt_laser/ui_data(mob/user)
+	var/list/data = list()
+	data["capacity"] = src.capacity
+	data["charge"] = src.charge
+	data["chargingEnabled"] = src.charging
+	data["excessPower"] = src.excess
+	data["gridLoad"] = src.terminal.powernet.load
+	data["inputLevel"] = src.chargelevel
+	data["inputMultiplier"] = src.input_multi
+	data["inputNumber"] = src.input_number
+	data["isCharging"] = src.is_charging
+	data["isFiring"] = src.firing
+	data["laserEnabled"] = src.online
+	data["lifetimeEarnings"] = src.lifetime_earnings
+	data["name"] = src.name
+	data["outputLevel"] = src.output
+	data["outputMultiplier"] = src.output_multi
+	data["outputNumber"] = src.output_number
+	data["totalGridPower"] = src.terminal.powernet.avail
```

Comment:
Consider `?.` here, if for whatever dumb reason it's missing a terminal.

Reply:
added